src/h3/mod.rs

```diff
@@ -26,4 +26,830 @@
+    pub num_placeholders: u64,
+    pub max_header_list_size: u64,
+    pub qpack_max_table_capacity: u64,
+    pub qpack_blocked_streams: u64,
+}
+
+impl H3Config {
+    #[allow(clippy::new_ret_no_self)]
+    pub fn new() -> Result<H3Config> {
+        Ok(H3Config {
+            root_dir: String::new(),
+            num_placeholders: 16,
+            max_header_list_size: 0,
+            qpack_max_table_capacity: 0,
+            qpack_blocked_streams: 0,
+        })
+    }
+
+    pub fn set_root_dir(&mut self, root_dir: &String) {
+        self.root_dir = String::clone(root_dir);
+    }
+
+    pub fn set_num_placeholders(&mut self, num_placeholders: u64) {
+        self.num_placeholders = num_placeholders;
+    }
+
+    pub fn set_max_header_list_size(&mut self, max_header_list_size: u64) {
+        self.max_header_list_size = max_header_list_size;
+    }
+
+    pub fn set_qpack_max_table_capacity(
+        &mut self, qpack_max_table_capacity: u64,
+    ) {
+        self.qpack_max_table_capacity = qpack_max_table_capacity;
+    }
+
+    pub fn set_qpacked_blocked_streams(&mut self, qpack_blocked_streams: u64) {
+        self.qpack_blocked_streams = qpack_blocked_streams;
+    }
+}
+
+/// An HTTP/3 connection.
+pub struct H3Connection {
+    root_dir: String,
+    is_server: bool,
+
+    num_placeholders: u64,
+    max_header_list_size: u64,
+    qpack_max_table_capacity: u64,
+    qpack_blocked_streams: u64,
+
+    peer_num_placeholders: std::option::Option<u64>,
+    peer_max_header_list_size: std::option::Option<u64>,
+    peer_qpack_max_table_capacity: std::option::Option<u64>,
+    peer_qpack_blocked_streams: std::option::Option<u64>,
+
+    control_stream_open: bool,
+    peer_control_stream_open: bool,
+
+    qpack_encoder: qpack::Encoder,
+    qpack_decoder: qpack::Decoder,
+
+    qpack_encoder_stream_open: bool,
+    peer_qpack_encoder_stream_open: bool,
+    qpack_decoder_stream_open: bool,
+    peer_qpack_decoder_stream_open: bool,
+}
+
+struct H3Util {}
+impl H3Util {
+    pub fn req_hdrs_to_qpack(
+        encoder: &mut qpack::Encoder, request: &http::Request<()>,
+    ) -> Vec<u8> {
+        let mut vec = vec![0u8; 65535];
+
+        let mut headers: Vec<qpack::Header> = Vec::new();
+
+        headers.push(qpack::Header::new(":method", request.method().as_str()));
+        headers.push(qpack::Header::new(
+            ":scheme",
+            request.uri().scheme_str().unwrap(),
+        ));
+        headers.push(qpack::Header::new(
+            ":authority",
+            request.uri().host().unwrap(),
+        ));
+        headers.push(qpack::Header::new(
+            ":path",
+            request.uri().path_and_query().unwrap().as_str(),
+        ));
+
+        for (key, value) in request.headers().iter() {
+            headers
+                .push(qpack::Header::new(key.as_str(), value.to_str().unwrap()));
+        }
+
+        let len = encoder.encode(&headers, &mut vec);
+
+        vec.truncate(len.unwrap());
```

Comment:
Must not use `unwrap()` as part of library code unless you are sure it's not going to crash (this applies to many other cases as well).

Reply:
Yup, need a full sweep of error handling approach